File: src/algorithms/CountingSort.cpp
#include "CountingSort.h"
#include <vector>
#include <algorithm>

std::string CountingSort::name() const {
    return "Counting Sort";
}

void CountingSort::sort(std::vector<int>& arr) {
    if (arr.empty()) return;

    int maxElement = *std::max_element(arr.begin(), arr.end());

    // Safety check: avoid unreasonable memory usage
    if (maxElement > 1'000'000) {
        // Fallback: do nothing (or could throw/log)
        return;
    }

    std::vector<int> count(maxElement + 1, 0);

    // Count occurrences
    for (int x : arr) {
        count[x]++;
    }

    // Reconstruct sorted array
    int index = 0;
    for (int i = 0; i <= maxElement; i++) {
        while (count[i]--) {
            arr[index++] = i;
        }
    }
}

File: java/algovista/algorithms/CountingSort.java
package algovista.algorithms;

public class CountingSort implements SortingAlgorithm {
    @Override
    public String name() {
        return "Counting Sort";
    }

    @Override
    public void sort(int[] arr) {
        if (arr.length == 0) {
            return;
        }

        int maxElement = arr[0];
        for (int value : arr) {
            if (value > maxElement) {
                maxElement = value;
            }
        }

        if (maxElement > 1_000_000) {
            return;
        }

        int[] count = new int[maxElement + 1];

        for (int value : arr) {
            count[value]++;
        }

        int index = 0;
        for (int i = 0; i <= maxElement; i++) {
            while (count[i]-- > 0) {
                arr[index++] = i;
            }
        }
    }
}
